actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,253 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+			continue
+		}
+
+		// TODO: is there an upper limit on number of deals in a sector (beside the practical limit imposed by sector sizes and min deal size)
+		// TODO Confirm the sector is actually in the stated deadline? This happens later when we try to update partition, but maybe do here first?
+
+		sectorInfo, err := sectors.MustGet(update.SectorID)
+		if err != nil {
+			rt.Log(rtt.INFO, "failed to get sector, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: this will change to "was originally dealless, and is currently no active deals"
+		if len(sectorInfo.DealIDs) != 0 {
+			rt.Log(rtt.INFO, "cannot update sector with deals, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: confirm sector is healthy
+
+		// TODO: What's the impact of the following flow:
+		// 	- activating the deals
+		// - failing to verify the proof (or some other failure later)
+		// - allowing the message to succeed anyway ("skipping" the sector)
+		// Does that leave the deal uncommitable (since state doesn't get reverted)? is it better to just fail the msg, revert state, and allow for a retry later?
+		code := rt.Send(
+			builtin.StorageMarketActorAddr,
+			builtin.MethodsMarket.ActivateDeals,
+			&market.ActivateDealsParams{
+				DealIDs:      update.Deals,
+				SectorExpiry: sectorInfo.Expiration,
+			},
+			abi.NewTokenAmount(0),
+			&builtin.Discard{},
+		)
+
+		if code != exitcode.Ok {
+			rt.Log(rtt.INFO, "failed to activate deals on sector %d, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		prevalidatedUpdates = append(prevalidatedUpdates, &updateAndSectorInfo{
+			update:     &update,
+			sectorInfo: sectorInfo,
+		})
+
+		sectorsDeals = append(sectorsDeals, market.SectorDeals{DealIDs: update.Deals, SectorExpiry: sectorInfo.Expiration})
+		sectorsDataSpec = append(sectorsDataSpec, &market.SectorDataSpec{
+			SectorType: sectorInfo.SealProof,
+			DealIDs:    update.Deals,
+		})
+	}
+
+	//// TODO: this method has some potential aborts -- are those all programmer error or is there scope to improve UX for bad input?
+	dealWeights := requestDealWeights(rt, sectorsDeals)
+
+	builtin.RequirePredicate(rt, len(dealWeights.Sectors) == len(prevalidatedUpdates), exitcode.ErrIllegalState,
+		"deal weight request returned %d records, expected %d", len(dealWeights.Sectors), len(prevalidatedUpdates))
+
+	// TODO: this method has some potential aborts -- are those all programmer error or is there scope to improve UX for bad input?
+	unsealedSectorCIDs := requestUnsealedSectorCIDs(rt, sectorsDataSpec...)
+
+	builtin.RequirePredicate(rt, len(unsealedSectorCIDs) == len(prevalidatedUpdates), exitcode.ErrIllegalState,
+		"unsealed sector cid request returned %d records, expected %d", len(unsealedSectorCIDs), len(prevalidatedUpdates))
+
+	succeededSectors := bitfield.New()
+	var st State
+	rt.StateTransaction(&st, func() {
+		for k, updateWithSectorInfo := range prevalidatedUpdates {
```

Comment:
Something to consider is doing updates by iterating over deadlines and partitions rather than updates.  During the validation loop we could group updates by deadline and partition in a map and then iterate over it here.  Again take a look at `ExtendSectorExpiration` for how this looks.  We could save batch updaters some deadline and partition ipld write gas this way.

Reply:
Yeah, I have a TODO in there for that, i think it's the next (and hopefully last) improvement-tweak needed for the MVP.